### DISCUSSION TITLE
Is there a way to access workspace files during feature execution?

### DISCUSSION THREAD

**Discussion by nikaro** [devcontainers/spec#404] [Q&A] [answered]
I am trying something like this:
```json
{
  "name": "exercice2-go",
  "image": "mcr.microsoft.com/devcontainers/go:1-1.21-bullseye",
  "features": {
    "ghcr.io/nikaro/features/ansible:1": {
      "dependencies": "-r ${containerWorkspaceFolder}/deploy/requirements.txt"
    }
  }
}
```

The feature uses the `dependencies` argument like this:
```sh
if [ -n "${DEPENDENCIES:-}" ]; then
  # ensure packages are separated by spaces
  DEPENDENCIES=$(printf '%s' "$DEPENDENCIES" | tr ',' ' ')
  "${VENV_PATH}/bin/pip" install $DEPENDENCIES
fi
```

And it fails with this error:
```
ERROR: Could not open requirements file: [Errno 2] No such file or directory: '/workspaces/Exercice_2/deploy/requirements.txt'
```

Is it possible to do something like that in some way?

**Comment by samruddhikhandale** [comment]
Yes, you can access workspace files during the execution of a Dev Container Feature. However, the timing of when the Feature scripts are run in relation to when the workspace files are mounted can affect this.

In a Dev Container, the workspace files are mounted into the container at the path specified by the `workspaceMount` property in `devcontainer.json` (or at `/workspaces/<folder-name>` by default if `workspaceMount` is not specified).

Feature scripts are run after the Docker image is built but before the workspace files are mounted. This means that if you try to access workspace files in a feature script, they might not be available yet.

To work around this, you can use a `postCreateCommand` or `postStartCommand` in `devcontainer.json` to run commands after the workspace files are mounted. These commands are run every time the container is created or started, respectively. I believe, you could also make use of [Feature's lifecycle hooks](https://containers.dev/implementors/features/#lifecycle-hooks). 

Looping in @joshspicer, feel free to correct my understanding if necessary.

**Reply by nikaro** [reply]
Thank for your answer!

Using `postCreateCommand` (or similar) would defeat the purpose of the feature, because otherwise i could do everything i want through `postCreateCommand` with arbitrary commands.

AFAIK, [feature's lifecycle hooks](https://containers.dev/implementors/features/#lifecycle-hooks) cannot access features arguments (`dependencies` in my example)?

**Reply by joshspicer** [reply, marked as answer]
That's true that lifecycle hooks don't have access to Feature arguments - that sounds like a helpful feature request, and would be useful in this scenario. 

One way I can think to do this today is similar to how [the `git-lfs` Feature works](https://github.com/devcontainers/features/blob/main/src/git-lfs/install.sh#L191-L217),  writing the lifecycle hook script in the `install.sh` and injecting the relevant Feature's options.  That script is then run [as a postCreate hook](https://github.com/devcontainers/features/blob/main/src/git-lfs/devcontainer-feature.json#L23) once the filesystem is mounted.

Note the usage of both `EOF` and `'EOF'` to control when variables are substituted or not (https://unix.stackexchange.com/questions/138418/passing-a-variable-to-a-bash-script-that-uses-eof-and-considers-the-variable-a)

**Reply by joshspicer** [reply]
> Using postCreateCommand (or similar) would defeat the purpose of the feature, because otherwise i could do everything i want through postCreateCommand with arbitrary commands.



Also, I don't think this is defeating the purpose of the Feature. You still need to install `ansible`, which is a natural thing to do in the Feature's `install.sh` script.  Tying the lifecycle hook to the Feature also makes this functionality more portable (people can just add your Feature instead of copy/pasting config into their repo!)

**Reply by nikaro** [reply]
Yes, the `postCreateCommand` at the feature's lifecycle hook would do the trick, but not at the `.devcontainer.json` level (or at least it would feel like a hack/workaround to me).

I will try the trick of the `git-lfs` feature, i think it should work for my use case.

**Reply by nikaro** [reply]
I confirm that it worked. Thank you @samruddhikhandale and @joshspicer for your help 🙏 

Here is the feature for the record: https://github.com/nikaro/features/tree/main/src/ansible

**Reply by joshspicer** [reply]
Thanks for sharing!
